sstables/sstables.cc

```diff
@@ -141,6 +141,36 @@ future<> sstable::rename_new_sstable_component_file(sstring from_name, sstring t
     });
 }
 
+class sstable::filesystem_storage final : public sstable::storage {
+    sstring dir;
+    std::optional<sstring> temp_dir; // Valid while the sstable is being created, until sealed
+
+private:
+    future<> check_create_links_replay(const sstable& sst, const sstring& dst_dir, generation_type dst_gen, const std::vector<std::pair<sstables::component_type, sstring>>& comps) const;
+    future<> remove_temp_dir();
+    virtual future<> create_links(const sstable& sst, const sstring& dir) const override;
+    future<> create_links_common(const sstable& sst, sstring dst_dir, generation_type dst_gen, mark_for_removal mark_for_removal) const;
+    future<> touch_temp_dir(const sstable& sst);
+    future<> move(const sstable& sst, sstring new_dir, generation_type generation, delayed_commit_changes* delay) override;
+
+    virtual void change_dir_for_test(sstring nd) override {
+        dir = std::move(nd);
+    }
+
```

Comment:
@xemul out of curiosity, why can't we reuse change_state() to achieve change of directory of tests?

Reply:
Because change_state() moves the whole sstable into new location. The change_dir_for_tests just forces the sstable to point at another location whatever files are there. I cannot yet wrap my head why sstable_test.cc::do_write_sst() need this hackery (it's the only user), but probably it won't like if files migrate between directories :)

Reply:
It's indeed some sinister hackery to just rewrite some of the sstable components elsewhere :-) 

Anyway, we can revisit it later. Not much work to get rid of it.